cirq-core/cirq/sim/act_on_args.py

```diff
@@ -113,14 +113,14 @@ def _perform_measurement(self, qubits: Sequence['cirq.Qid']) -> List[int]:
         """Child classes that perform measurements should implement this with
         the implementation."""
 
-    def copy(self: TSelf) -> TSelf:
+    def copy(self: TSelf, with_buffer: bool = True) -> TSelf:
         """Creates a copy of the object."""
         args = copy.copy(self)
-        self._on_copy(args)
+        self._on_copy(args, with_buffer)
         args._log_of_measurement_results = self.log_of_measurement_results.copy()
         return args
 
-    def _on_copy(self: TSelf, args: TSelf):
+    def _on_copy(self: TSelf, args: TSelf, with_buffer: bool = True):
```

Comment:
Since this is meant to be overridden, I'm not sure we can add a parameter to it without breaking external implementations of ActOnArgs, unless there's some Python magic that allows it. We may have to add a separate `_on_copy_reusing_buffer` (which has a default implementation calling `_on_copy()`), and have `copy` call the appropriate one according to the input.

Reply:
Also I think I prefer the arg name `reuse_buffer` with default value False.

Reply:
I agree that adding a parameter will break external implementations of `ActOnArgs`, but adding a new method has drawbacks, too. The most challenging part is to connect the new method with `copy`. I noticed that you have updated the issue. So do you now support the way of adding a parameter to `copy`?

Reply:
Yeah, that's a tough one. Technically changing `OperationTarget.copy`'s signature is a breaking change too, though I never intended `OperationTarget` to be implemented directly by third parties.

I'd probably go about it the way you've done so far. But in `ActOnArgs.copy`, I'd suggest adding an `if 'reuse_buffer' in inspect.signature(_on_copy)` check, then either pass or don't pass the argument accordingly. Then also if `reuse_buffer` is not in the signature, emit a deprecation warning. We could do the same thing when calling `a.copy(reuse_buffer)` in ActOnArgsContainer if we want to be especially cautious. (Sorry about changing the issue mid-flight, I didn't think anybody would be actively working on it).

@95-martin-orion do you have an opinion here?

Reply:
@daxfohl I am just trying to learn new things, and I have learned a lot so far! Thanks for your patient explanations. I will look into the suggestions you list here and in the issue.